### PR TITLE
 sys/result_output: Initial result output abstraction

### LINKED ISSUE #15037: [RFC] abstraction layer for test commands
#### Description

To help with writing tests conforming to a convention I suggest we add an abstraction layer for use with writing tests.

The concept is instead of printing results without structure (though obvious for humans) we use an abstraction for the results. Not only should this simplify writing tests but also allow easier automation of testing. It allows changing format without changing test logic, for example, a serializer can replace the printf messages to save size.

[Here](https://github.com/RIOT-OS/RIOT/blob/9359b97b21473b832eadc4cdab4b8834676081dd/sys/include/test_utils/result_output.h) is an example API.

The results of calling the functions would look something like:
```
# turo_data_int(TH_DEV, 12)
data: 12

# turo_data_bool(TH_DEV, true)
data: True

# turo_data_string(TH_DEV, "foo")
data: "foo"

# turo_data_array(TH_DEV, [1, -1, 256, 0xFFFF, 0xFFFFFFFF], INT32_TYPE, 5);
data: [1,-1,256,65536,-1]

# turo_data_string_object(TH_DEV, "foo", "bar")
data: {"foo": "bar"}

# turo_data_string_object(TH_DEV, "foo", 42)
data: {"foo": 42}

#turo_result_success()
result: Success

#turo_result_error()
result: Error

# turo_result_error_with_code(-ENODEV)
error_code: -12
result: Error
```

Note, for futureproofing I have a `int turo` for different contexts.

The format is chosen since it is very readable and easy enough for the firmware to create and for a higher level language to parse (I would rather have the parsing logic on the high level rather than in the firmware).
It is also doesn't require large buffers or states. The data printout is versatile. It is similar to what we have now for some applications. It is very human readable. It also allows debug messages in-between without interfering with the parsing.
I intentionally leave out all the different types of primitives, int32_t` covers many cases and this can get extended as needed.

Here are some questions I would like feedback on:
1. Should this be a different module or combined with the `fmt` library?
1. How is the naming of the module (test_helpers) and th namespacing?
1. Do I really need to specify each function as *_print_*?
1. For the array print should I just have a function for each type or have an enum situation?

Feedback is appreciated!

The long term goal will be to define a testing convention and integrate a riotctrl module and easily automate shell/api based testing.

### Useful links
Discussed in #10624

### PR DESCRIPTION
**Pull request by MrKevinWeiss**
### Contribution description

This provides an abstraction and implementation of simple data print and result functions to be used to standardize RIOT test firmware.

### Testing procedure

run `make flash term -C tests/periph_cpuid` and check the output.

### Issues/PRs references

Fixes #15037

### CODE REVIEW

**Comment by MrKevinWeiss** [comment]
Updated to full json output as per conversation with @kaspar030 since it will work better with murdock metrics

**Comment by MrKevinWeiss** [comment]
Updated `tests/bench_msg_pingpong` to use turo
```
2020-10-16 11:58:53,897 # START
2020-10-16 11:58:53,904 # main(): This is RIOT! (Version: 2020.10-devel-1992-gc7e3a-pr/testhelpers)
2020-10-16 11:58:54,906 # {"data": {"n":100397}}
2020-10-16 11:58:54,908 # {"data": {"ticks":717}}
2020-10-16 11:58:54,910 # {"result": "Success"}
```

**Comment by MrKevinWeiss** [comment]
Updated `tests/periph_cpuid`

```
2020-10-16 11:48:47,764 # START
2020-10-16 11:48:47,771 # main(): This is RIOT! (Version: 2020.10-devel-1992-g7668-pr/testhelpers)
2020-10-16 11:48:47,773 # Test for the CPUID driver
2020-10-16 11:48:47,778 # This test is reading out the CPUID of the platforms CPU
2020-10-16 11:48:47,778 # 
2020-10-16 11:48:47,779 # CPUID_LEN: 12
2020-10-16 11:48:47,784 # {"data": [48,255,108,6,78,84,52,56,17,40,16,67]}
2020-10-16 11:48:47,786 # {"result": "Success"}
```

**Comment by MrKevinWeiss** [comment]
I will run murdock to test metrics

**Comment by MrKevinWeiss** [comment]
This does seem to increase build size... I thought I could make some gains by using the `fmt` library...  Darn.

**Comment by kaspar030** [comment]
> 2020-10-16 11:58:54,906 # {"data": {"n":100397}}
> 2020-10-16 11:58:54,908 # {"data": {"ticks":717}}

why not just `{"n":100397}`?

**Comment by MrKevinWeiss** [comment]
I wanted to somehow namespace data.

**Comment by kaspar030** [comment]
> This does seem to increase build size... I thought I could make some gains by using the `fmt` library... Darn.

yeah, all stdio code is still used (just by "Hello, this is RIOT!")...

**Comment by MrKevinWeiss** [comment]
> yeah, all stdio code is still used (just by "Hello, this is RIOT!")...

gross, oh well, that is optimization (probably also used in the interactive sync.

File: sys/test_utils/result_output/result_output.c [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+#include "assert.h"
+#include "fmt.h"
+
+#include "test_utils/result_output.h"
+
+void turo_data_int(int turo, int32_t data)
+{
+    (void)turo;
+    print_str("{\"data\": ");
```
Comment (kaspar030) [inline sys/test_utils/result_output/result_output.c:31]: should we add a space here and there, so it looks a bit more pretty printed? e.g., here after "}"

Comment (kaspar030) [reply sys/test_utils/result_output/result_output.c:31]: we can do that in a folo

Comment (MrKevinWeiss) [reply sys/test_utils/result_output/result_output.c:31]: Ok, I will push after mrudock is done.

Comment (MrKevinWeiss) [reply sys/test_utils/result_output/result_output.c]: Done

**Comment by MrKevinWeiss** [comment]
> yeah, all stdio code is still used (just by "Hello, this is RIOT!")...

I guess #13710 will help with that!

**Comment by MrKevinWeiss** [comment]
@fjmolinas & @kaspar030 
if someone has time I would like this (or similar functionality) to move forward.

**Comment by MrKevinWeiss** [comment]
@kaspar030 ping,

There is an issues with this:
The tests written with this will look less and less like examples due to using this module instead of printf...

If we say the convenience of this is work that (and you should be able to replace these with just a printf) then we can move forward and even make a `riotctrl` parse that can interpret this.

**Comment by MrKevinWeiss** [comment]
After talking with @leandrolanzieri and @jia200x it seems that we think the best way is to have a way to open a result container chronologically print the data from the results and close it (ideally with an exit status similar to linux).
We can always do helper functions that takes care of the opening and closing so for simple tasks it won't be a burden.

Structure can be enforced with the CI for example, every command would have an exit status.

Adding a generic parser to `riotctrl` should be easy.

The one thing that wasn't agreed on is to allow unformatted data from `DEBUG` prints or generally from a printf as part of the output, the parser could somehow ignore that.

I will survey some serializer APIs to see what would be needed and provide something a bit more complete, and maybe a test that can show the capabilities a bit better.

**Comment by MrKevinWeiss** [comment]
closed with #15950

**Pull request closed without merging**